Backed Resource Dictionary with TwoWay x:Bind, nullables and WarningsAsErrors generates invalid code

### Current behavior

Backed resource dictionaries that use `{x:Bind SomeProp, Mode=TwoWay}`, referencing a nullable property in projects with "treat nullability warnings as errors" via `<WarningsAsErrors>nullable</WarningsAsErrors>` and nullables enabled will fail to compile, citing this error:

> error CS8669: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context. Auto-generated code requires an explicit '#nullable' directive in source.

![image](https://user-images.githubusercontent.com/9384894/138361099-ed2a96d5-7c60-4495-b32d-fdc2b5d4edb0.png)


### Expected behavior

No compile time errors.

### How to reproduce it (as minimally and precisely as possible)

1. Add this to your `.csproj`:
```xml
<Nullable>enable</Nullable>
<LangVersion>8.0</LangVersion>
<WarningsAsErrors>nullable</WarningsAsErrors>
```

2. Create a class with a nullable property
3. Create a backed resource dictionary with a DataTemplate
4. Add `{x:Bind YourProperty, Mode=TwoWay}` to the template
5. Attempt to compile

Premade project here:
[UnoNullableContextGeneratedBug.zip](https://github.com/unoplatform/uno/files/7393047/UnoNullableContextGeneratedBug.zip)

The following was tested:
- UWP project compiles
- WASM project does not compile



### Workaround

Don't use `Mode=TwoWay`, or remove `WarningsAsErrors` from the affected projects.

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia, Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia, Uno.UI.RemoteControl / Uno.WinUI.RemoteControl

### NuGet package version(s)

Uno.UI 3.10.11
Uno.UI.WebAssembly 3.10.11
Uno.UI.RemoteControl 3.10.11

### Affected platforms

WebAssembly

### IDE

Visual Studio 2019

### IDE version

16.11.4

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

--- a/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
+++ b/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
@@ -3875,7 +3875,7 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 						{
 							if (!string.IsNullOrWhiteSpace(rawBindBack))
 							{
-								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0], GetType(dataType));
+								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0], GetType(dataType)).ToDisplayString(NullableFlowState.None);
 								return $"(___ctx, __value) => {{ if(___ctx is {dataType} ___tctx) {{ ___tctx.{rawBindBack}(({propertyType})__value); }} }}";
 							}
 							else
@@ -3887,8 +3887,8 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 						{
 							if (propertyPaths.properties.Length == 1)
 							{
-								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0], GetType(dataType));
-								return $"(___ctx, __value) => {{ if(___ctx is {dataType} ___tctx) {{ {contextFunction} = ({targetPropertyType})global::Windows.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(typeof({targetPropertyType.ToDisplayString(NullableFlowState.None)}), __value); }} }}";
+								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0], GetType(dataType)).ToDisplayString(NullableFlowState.None);
+								return $"(___ctx, __value) => {{ if(___ctx is {dataType} ___tctx) {{ {contextFunction} = ({targetPropertyType})global::Windows.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(typeof({targetPropertyType}), __value); }} }}";
 							}
 							else
 							{
@@ -3928,10 +3928,10 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 						{
 							if (propertyPaths.properties.Length == 1)
 							{
-								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0]);
+								var targetPropertyType = GetXBindPropertyPathType(propertyPaths.properties[0]).ToDisplayString(NullableFlowState.None);
 								return $"(___ctx, __value) => {{ " +
 									$"if(___ctx is global::{_className.ns + "." + _className.className} ___tctx) " +
-									$"{rawFunction} = ({targetPropertyType})global::Windows.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(typeof({targetPropertyType.ToDisplayString(NullableFlowState.None)}), __value);" +
+									$"{rawFunction} = ({targetPropertyType})global::Windows.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(typeof({targetPropertyType}), __value);" +
 									$" }}";
 							}
 							else
@@ -4105,7 +4105,7 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 				{
 					// The target property implements IConvertible, therefore
 					// cast ProvideValue() using Convert.ChangeType
-					var targetTypeDisplay = propertyType.ToDisplayString();
+					var targetTypeDisplay = propertyType.ToDisplayString(NullableFlowState.None);
 					var targetType = $"typeof({targetTypeDisplay})";
 
 					// It's important to cast to string before performing the conversion

--- a/src/SourceGenerators/XamlGenerationTests/TestNullable_ResourceDictionary.xaml
+++ b/src/SourceGenerators/XamlGenerationTests/TestNullable_ResourceDictionary.xaml
@@ -1,0 +1,17 @@
+﻿<ResourceDictionary
+    x:Class="UnoNullableContextGeneratedBug.TestTemplate"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:UnoNullableContextGeneratedBug">
+
+    <DataTemplate x:Key="TestDataTemplate" x:DataType="local:TestViewModel">
+        <StackPanel>
+            <!--This works-->
+            <TextBlock Text="{x:Bind ObjectInstance, Mode=OneWay}"/>
+        
+            <!-- This generates code with a nullable error at compile time -->
+            <TextBlock Text="{x:Bind ObjectInstance, Mode=TwoWay}"/>
+        </StackPanel>
+    </DataTemplate>
+
+</ResourceDictionary>

--- a/src/SourceGenerators/XamlGenerationTests/TestNullable_ViewModel.cs
+++ b/src/SourceGenerators/XamlGenerationTests/TestNullable_ViewModel.cs
@@ -1,0 +1,24 @@
+﻿#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Text;
+
+namespace UnoNullableContextGeneratedBug
+{
+    public class TestViewModel
+    {
+        public TestViewModel()
+        {
+            ObjectInstance = new ThrowawayClass();
+        }
+
+        public ThrowawayClass? ObjectInstance { get; set; } = new ThrowawayClass();
+    }
+
+    public class ThrowawayClass
+    {
+    }
+}

--- a/src/SourceGenerators/XamlGenerationTests/XamlGenerationTests.csproj
+++ b/src/SourceGenerators/XamlGenerationTests/XamlGenerationTests.csproj
@@ -2,6 +2,9 @@
 	<PropertyGroup>
 		<TargetFrameworks>MonoAndroid11.0;xamarinios10;netstandard2.0</TargetFrameworks>
 		<TargetFrameworksCI>MonoAndroid11.0;MonoAndroid10.0;xamarinios10;netstandard2.0</TargetFrameworksCI>
+
+		<LangVersion>8.0</LangVersion>
+		<WarningsAsErrors>nullable</WarningsAsErrors>
 	</PropertyGroup>
 
 	<PropertyGroup>

